daemon/qrexec-client.c

```diff
@@ -482,22 +531,24 @@ static int handle_vchan_data(libvchan_t *vchan, struct buffer *stdin_buf,
 
 static void check_child_status(libvchan_t *vchan)
 {
-    pid_t pid;
-    int status;
-
-    pid = waitpid(local_pid, &status, WNOHANG);
-    if (pid < 0) {
-        perror("waitpid");
-        do_exit(1);
+    int status = 0;
+    if (local_pid >= 0) {
+        /* this is not a socket-based service */
+        pid_t pid = waitpid(local_pid, &status, WNOHANG);
+        if (pid < 0) {
+            perror("waitpid");
+            close_vchan_and_exit(1, vchan);
+        }
+        if (pid == 0 || !WIFEXITED(status))
+            return;
+        status = WEXITSTATUS(status);
     }
-    if (pid == 0 || !WIFEXITED(status))
-        return;
     if (is_service)
-        send_exit_code(vchan, WEXITSTATUS(status));
-    do_exit(status);
+        send_exit_code(vchan, status);
+    close_vchan_and_exit(1, vchan);
```

Comment:
@DemiMarie The local cmd handling issue is here: hardcoded 1 instead of `status` as it was previously.

@pwmarcz I think your test may be wrong, if I fix it to `status`, then the exit code is `44`, not `0`. On the other hand, I'm not really sure if `qrexec-client` exit code should represent _local_ process exit code, instead of the remote one. In case of no local command, it should definitely be remote process exit code. But if there is local command logically it should receive remote exit code - with an exception that we have no way of doing that... What do you think how it _should_ behave?

Reply:
And to complete the picture - if `qrexec-client` should exit with the remote process exit code even if local command is used, then both code and test are wrong - code does not wait for remote exit code and test doesn't send one.

Reply:
My test reflected the status quo on master - currently if the local command exits first, we always exit with code 0, and if the remote command exits first, we return its code. I agree that sounds wrong.

I also think returning the remote exit code sounds good. We cannot do that in the case where we just connect two remote domains, but in this case (with both local cmd and remote cmd) probably it makes sense to wait for the remote end.